src/devenv/cli.py

```diff
@@ -0,0 +1,689 @@
+import os
+import shlex
+import shutil
+import signal
+import subprocess
+import tempfile
+import time
+import re
+import sys
+from filelock import FileLock
+from contextlib import suppress
+from pathlib import Path
+
+import pkgutil
+import json
+
+import click
+import terminaltables
+import strictyaml
+
+
+from .yaml import validate_and_parse_yaml, read_yaml, write_yaml, schema
+from .log import log, log_task
+
+
+NIX_FLAGS = [
+    "--show-trace",
+    "--extra-experimental-features",
+    "nix-command",
+    "--extra-experimental-features",
+    "flakes",
+    "--option",
+    "warn-dirty",
+    "false",
+]
+FILE = pkgutil.get_loader(__package__).load_module(__package__).__file__
+if 'site-packages' in FILE:
+    SRC_DIR = Path(FILE, '..', '..', 'src')
+else:
+    SRC_DIR = Path(FILE, '..', '..')
+MODULES_DIR = (SRC_DIR / 'modules').resolve()
+FLAKE_FILE_TEMPL = os.path.join(MODULES_DIR, "flake.tmpl.nix")
+FLAKE_FILE = ".devenv.flake.nix"
+FLAKE_LOCK = "devenv.lock"
+
+# define system like x86_64-linux
+SYSTEM = os.uname().machine.lower().replace("arm", "aarch") + "-" + os.uname().sysname.lower()
+
+def run_nix(command: str, skip_exc_wrapping=False, replace_shell=False) -> str:
+    ctx = click.get_current_context()
+    nix_flags = ctx.obj['nix_flags']
+    flags = " ".join(NIX_FLAGS) + " " + " ".join(nix_flags)
+    command_flags = " ".join(ctx.obj['command_flags'])
+    return run_command(f"nix {flags} {command} {command_flags}",
+                       skip_exc_wrapping=skip_exc_wrapping,
+                       replace_shell=replace_shell)
+
+def run_command(command: str, disable_stderr=False, skip_exc_wrapping=False, replace_shell=False) -> str:
```

Comment:
I would default to a list here actually to avoid having to escape shell arguments.

Reply:
Try https://amoffat.github.io/sh/
Amazing library to make shell calls.

Reply:
sh is a full-fledged subprocess replacement for Python 2.6 - 3.8, PyPy and PyPy3 that allows you to call any program as if it were a function: